src/hotspot/share/gc/g1/g1CollectionSetCandidates.hpp

```diff
@@ -29,72 +29,208 @@
 #include "gc/shared/workerThread.hpp"
 #include "memory/allocation.hpp"
 #include "runtime/globals.hpp"
+#include "utilities/bitMap.hpp"
+#include "utilities/growableArray.hpp"
 
+class G1CollectionCandidateList;
+class G1CollectionSetCandidates;
 class HeapRegion;
 class HeapRegionClosure;
 
-// Set of collection set candidates, i.e. all old gen regions we consider worth
-// collecting in the remainder of the current mixed phase. Regions are sorted by decreasing
-// gc efficiency.
-// Maintains a cursor into the list that specifies the next collection set candidate
-// to put into the current collection set.
-class G1CollectionSetCandidates : public CHeapObj<mtGC> {
-  HeapRegion** _regions;
-  uint _num_regions; // Total number of regions in the collection set candidate set.
+using G1CollectionSetRegionListIterator = GrowableArrayIterator<HeapRegion*>;
 
-  // The sum of bytes that can be reclaimed in the remaining set of collection
-  // set candidates.
-  size_t _remaining_reclaimable_bytes;
-  // The index of the next candidate old region to be considered for
-  // addition to the current collection set.
-  uint _front_idx;
+// A set of HeapRegion*.
+class G1CollectionSetRegionList {
```

Comment:
Now that this is just a region-list, maybe drop the "CollectionSet" part?

Reply:
I would like to keep the name as is and avoid generalizations that are unnecessary at this time. If there is additional use for it, we can always factor it out.

Reply:
It's mostly to avoid confusion. The two even have the same length...

```
G1CollectionCandidateList
G1CollectionSetRegionList
```